drivers/pcie/host/pcie.c

```diff
@@ -46,26 +46,51 @@ void pcie_set_cmd(pcie_bdf_t bdf, uint32_t bits, bool on)
 	pcie_conf_write(bdf, PCIE_CONF_CMDSTAT, cmdstat);
 }
 
-uintptr_t pcie_get_mbar(pcie_bdf_t bdf, unsigned int index)
+struct pci_mbar pcie_get_mbar(pcie_bdf_t bdf, unsigned int index)
 {
 	uint32_t reg, bar;
-	uintptr_t addr = PCIE_CONF_BAR_NONE;
-
-	reg = PCIE_CONF_BAR0;
-	for (bar = 0; bar < index && reg <= PCIE_CONF_BAR5; bar++) {
-		if (PCIE_CONF_BAR_64(pcie_conf_read(bdf, reg++))) {
-			reg++;
+	struct pci_mbar mbar = {0};
+
+	bar = 0;
+	for (reg = PCIE_CONF_BAR0; reg <= PCIE_CONF_BAR5; reg++) {
+		uint32_t addr = pcie_conf_read(bdf, reg);
+		if (addr != 0xFFFFFFFFU) {
+			if (bar == index) {
+				break;
+			}
+			bar++;
+			if (PCIE_CONF_BAR_64(addr)) {
+				reg++;
+			}
 		}
 	}
 
 	if (bar == index) {
-		addr = pcie_conf_read(bdf, reg++);
-		if (IS_ENABLED(CONFIG_64BIT) && PCIE_CONF_BAR_64(addr)) {
-			addr |= ((uint64_t)pcie_conf_read(bdf, reg)) << 32;
+		mbar.phys_addr = pcie_conf_read(bdf, reg);
+		pcie_conf_write(bdf, reg, 0xFFFFFFFF);
+		mbar.size = pcie_conf_read(bdf, reg);
+		pcie_conf_write(bdf, reg, (uint32_t)mbar.phys_addr);
+
+		if (IS_ENABLED(CONFIG_64BIT) && PCIE_CONF_BAR_64(mbar.phys_addr)) {
+			reg++;
+			mbar.phys_addr |= ((uint64_t)pcie_conf_read(bdf, reg)) << 32;
+			pcie_conf_write(bdf, reg, 0xFFFFFFFF);
+			mbar.size |= ((uint64_t)pcie_conf_read(bdf, reg)) << 32;
+			pcie_conf_write(bdf, reg, (uint32_t)((uint64_t)mbar.phys_addr >> 32));
+		}
+
+		mbar.size = PCIE_CONF_BAR_MEM(mbar.phys_addr) ? PCIE_CONF_BAR_MEM_ADDR(mbar.size) : PCIE_CONF_BAR_IO_ADDR(mbar.size);
+
+		if (!mbar.size) {
+			mbar.phys_addr = PCIE_CONF_BAR_NONE;
+		} else {
+			mbar.size &= ~(mbar.size-1);
+			mbar.flags = PCIE_CONF_BAR_MEM(mbar.phys_addr) ? mbar.phys_addr & 0xFUL : mbar.phys_addr & 0x3U;
```

Comment:
have macros for getting the flags instead of these "mbar.phys_addr & ... "
However, see below in the struct pcie_mbar declaration, I don't think we need these anyway